src/tree/updater_gpu_hist.cu

```diff
@@ -1028,21 +1033,20 @@ class GPUHistMakerSpecialised{
   }
 
   void AllReduceHist(int nidx) {
-    if (shards_.size() == 1) return;
-
-    monitor_.Start("AllReduce");
-    dh::ExecuteIndexShards(
-        &shards_,
-        [&](int idx, std::unique_ptr<DeviceShard<GradientSumT>>& shard) {
-          auto d_node_hist = shard->hist.GetNodeHistogram(nidx).data();
-          reducer_.AllReduceSum(
-              dist_.Devices().Index(shard->device_id_),
-              reinterpret_cast<typename GradientSumT::ValueT*>(d_node_hist),
-              reinterpret_cast<typename GradientSumT::ValueT*>(d_node_hist),
-              n_bins_ * (sizeof(GradientSumT) /
-                         sizeof(typename GradientSumT::ValueT)));
-        });
-    monitor_.Stop("AllReduce");
+    dh::safe_cuda(cudaDeviceSynchronize());
```

Comment:
This has changed the behaviour of the single machine code. The existing code uses one thread per GPU to launch the reduction, your code changes to use the group API and a single thread for all GPUs.